File: src/query/types/issuers.ts
import { Coin } from "@/query/types/coins.ts";

export interface Issuer {
  id: number;
  name: string;
  continent: string;
  start_year: number;
  end_year: number | null;
  flag: string;
  predecessors?: Issuer[];
  descendants?: Issuer[];
  issued_coins?: Coin[];
  created_at: string;
}

export interface IssuerDisplay {
  id: number;
  name: string;
  start_year: number;
  end_year: number | null;
  flag: string;
}

export interface PaginatedIssuers {
  items: Issuer[];
  total: number;
}

export interface ListIssuersRequest {
  page?: number;
  pageSize?: number;
  search?: string | null;
}

export interface GetIssuerRequest {
  id: number;
  limit?: number;
}

File: src-tauri/src/types/issuers.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Issuer {
    pub id: i32,
    pub name: String,
    pub continent: Option<String>,
    pub start_year: Option<i32>,
    pub end_year: Option<i32>,
    pub flag: String,
    #[serde(default)]
    pub predecessors: Option<Vec<Issuer>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub descendants: Option<Vec<Issuer>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub issued_coins: Option<Vec<crate::types::coins::Coin>>,
    pub created_at: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct IssuerDisplay {
    pub id: i32,
    pub name: String,
    pub start_year: Option<i32>,
    pub end_year: Option<i32>,
    pub flag: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct PaginatedIssuersResponse {
    pub items: Vec<Issuer>,
    pub total: i64,
}
